src/Formats/ProtobufSchemas.cpp

```diff
@@ -33,22 +35,36 @@ class ProtobufSchemas::ImporterWithSourceTree : public google::protobuf::compile
 
     const google::protobuf::Descriptor * import(const String & schema_path, const String & message_name)
     {
-        // Search the message type among already imported ones.
-        const auto * descriptor = importer.pool()->FindMessageTypeByName(message_name);
-        if (descriptor)
-            return descriptor;
```

Comment:
Why did you remove it?

Reply:
The previous logic first probed a cache for the descriptor of an already imported schema (l. 37-39). If it could not be found in the cache, the import was done manually (l. 41-42).

However, importer.Import() probes the same cache as well (*). This means the manual probing was redundant.

(*)

Reply:
Not really, here we search in cache by message name and in importer.Import() we search in cache by schema path. It prevents cases when we try to work with different schemas with same message names. I built ClickHouse server from this PR and made some tests, I was able to get segfault on server stop because of this change:
File simple_schema.proto:
```
syntax = "proto3";

 message MessageType {
    uint32 colA = 1;
    string colB = 2;
}
```
File simple_schema2.proto is the copy of simple_schema.proto. 

```bash
$ clickhouse-client -q "select number as colA, 'String' as colB from numbers(10) format Protobuf settings format_schema='simple_schema:MessageType'" > data.bin
$ clickhouse-client -q "select * from file('data.bin', 'Protobuf', 'colA UInt32, colB String') settings format_schema='simple_schema:MessageType'" > /dev/null
$ clickhouse-client -q "select * from file('data.bin', 'Protobuf', 'colA UInt32, colB String') settings format_schema='simple_schema2:MessageType'"
Received exception from server (version 22.3.1):
Code: 434. DB::Exception: Received from localhost:9000. DB::Exception: Cannot parse 'simple_schema2.proto' file, found an error at line 3, column 11, "MessageType.colA" is already defined in file "simple_schema.proto".: While executing File. (CANNOT_PARSE_PROTOBUF_SCHEMA)
(query: select * from file('data.bin', 'Protobuf', 'colA UInt32, colB String') settings format_schema='simple_schema2:MessageType')
```

And if I try to stop ClickHouse server I will see this:
```
[libprotobuf FATAL ../contrib/protobuf/src/google/protobuf/descriptor.cc:810] CHECK failed: checkpoints_.empty():
Segmentation fault (core dumped)
```

Reply:
That is a very good point. The semantics were changed and I will roll my change back to get the original behavior. Thanks!

I tried to reproduce the error/segfault with your repro but unfortunately failed. (More specifically, both SELECT * queries aborted with an error that the schemafile could not be found. I thought the user_files/ path may be misconfigured and provided a custom config file to clickhouse-server (via -C) to configure user_files but that somehow made the server access data in /var/lib/clickhouse/ instead of data in my home directory. On Mac, /var/ can unfortunately not be made permissive easily ... But perhaps I just need to gain more experience with the configuration since I am pretty new to the project.)

On a related note, I suppose that (generally speaking) cache lookups may in some cases be counterintuitive. E.g., a user could wonder why modifications of a schemafile are not getting picked up by ClickHouse which simply continues to use the old version from the cache found by the message name. Unfortunately, I also could not create a scenario which shows that because somehow all queries in my local setup are fully executed inside clickhouse-client (instead of clickhouse-server) so that the cache is wiped between queries.